### PR TITLE
Support unix feature on OpenBSD (utmpx not supported)

### PR DESCRIPTION
**Pull request by lcheylus**
- "unix" feature requires `utmpx` support for `pinky`, `users`, `uptime` and `who` tools.
- `utmpx` is not supported on OpenBSD => add conditional compilation for these 4 tools to disable it and display "unsupported command on OpenBSD" when called.

Fixes uutils/coreutils#5596

---

- Build OK with feature "unix" on OpenBSD-current/amd64 with Rust 1.74
```
$ cargo build --release --features unix
(...)

$ target/release/coreutils
target/release/coreutils
coreutils 0.0.23 (multi-call binary)

Usage: coreutils [function [arguments...]]

Currently defined functions:

    [, arch, b2sum, b3sum, base32, base64, basename, basenc, cat, chgrp, chmod, chown, chroot,
    cksum, comm, cp, csplit, cut, date, dd, df, dir, dircolors, dirname, du, echo, env, expand,
    expr, factor, false, fmt, fold, groups, hashsum, head, hostid, hostname, id, install, join,
    kill, link, ln, logname, ls, md5sum, mkdir, mkfifo, mknod, mktemp, more, mv, nice, nl,
    nohup, nproc, numfmt, od, paste, pathchk, pinky, pr, printenv, printf, ptx, pwd, readlink,
    realpath, rm, rmdir, seq, sha1sum, sha224sum, sha256sum, sha3-224sum, sha3-256sum, sha3-
    384sum, sha3-512sum, sha384sum, sha3sum, sha512sum, shake128sum, shake256sum, shred, shuf,
    sleep, sort, split, stat, stdbuf, stty, sum, sync, tac, tail, tee, test, timeout, touch,
    tr, true, truncate, tsort, tty, uname, unexpand, uniq, unlink, uptime, users, vdir, wc, who,
    whoami, yes

$ target/release/coreutils uname -a
OpenBSD openbsd-dev.home.lan 7.4 GENERIC.MP#1471 amd64 OpenBSD

$ target/release/coreutils pinky
unsupported
$ target/release/coreutils uptime
unsupported
$ target/release/coreutils users
unsupported
$ target/release/coreutils who
unsupported
```

- No regression on Linux : build OK with Rust v1.74.0 on Debian testing/amd64.

- Tests on OpenBSD with feature "unix"
```
$ cargo test --tests --features unix
(...)
failures:
    test_cat::test_error_loop
    test_chown::test_chown_no_change_to_group
    test_chown::test_chown_no_change_to_user_group
    test_chown::test_chown_only_group
    test_chown::test_chown_only_group_id
    test_chown::test_chown_owner_group
    test_chown::test_chown_owner_group_id
    test_chown::test_chown_owner_group_mix
    test_chown::test_chown_various_input
    test_cp::test_copy_contents_fifo
    test_cp::test_cp_parents_2_dirs
    test_cp::test_cp_preserve_xattr
    test_cp::test_no_preserve_mode
    test_cp::test_preserve_hardlink_attributes_in_directory
    test_cp::test_preserve_mode
    test_du::test_du_basics_subdir
    test_du::test_du_d_flag
    test_du::test_du_dereference
    test_du::test_du_hard_link
    test_du::test_du_no_dereference
    test_du::test_du_no_permission
    test_du::test_du_one_file_system
    test_du::test_du_soft_link
    test_du::test_du_symlink_multiple_fail
    test_du::test_du_threshold
    test_hostname::test_hostname_ip
    test_install::test_install_compare_option
    test_install::test_install_copy_then_compare_file
    test_ls::test_ls_allocation_size
    test_ls::test_ls_human_si
    test_ls::test_ls_long_total_size
    test_pinky::test_long_format
    test_pinky::test_long_format_multiple_users
    test_pinky::test_no_flag
    test_pinky::test_short_format_i
    test_pinky::test_short_format_q
    test_split::test_dev_zero
    test_split::test_filter_broken_pipe
    test_split::test_number_by_bytes_dev_zero
    test_stat::test_format_created_time
    test_stat::test_multi_files
    test_stat::test_normal_format
    test_stat::test_pipe_fifo
    test_stat::test_stdin_pipe_fifo1
    test_stat::test_stdin_pipe_fifo2
    test_stat::test_stdin_redirect
    test_stat::test_symlinks
    test_stdbuf::test_stdbuf_line_buffered_stdout
    test_stdbuf::test_stdbuf_trailing_var_arg
    test_stdbuf::test_stdbuf_unbuffered_stdout
    test_tail::test_follow_descriptor_vs_rename1
    test_tail::test_follow_descriptor_vs_rename2
    test_tail::test_follow_name_move1
    test_tail::test_follow_name_move2
    test_tail::test_follow_name_move_create1
    test_tail::test_follow_name_move_create2
    test_tail::test_follow_name_move_retry1
    test_tail::test_follow_name_move_retry2
    test_tail::test_follow_name_retry_headers
    test_tail::test_follow_when_files_are_pointing_to_same_relative_file_and_file_stays_same_size
    test_tail::test_retry3
    test_tail::test_retry4
    test_tail::test_retry5
    test_tail::test_retry7
    test_tail::test_retry8
    test_tail::test_stdin_redirect_dir
    test_test::test_file_is_sticky
    test_test::test_file_owned_by_egid
    test_touch::test_touch_changes_time_of_file_in_stdout
    test_touch::test_touch_dash
    test_uptime::test_uptime
    test_uptime::test_uptime_since
    test_who::test_arg1_arg2
    test_who::test_boot
    test_who::test_dead
    test_who::test_login
    test_who::test_m
    test_who::test_process
    test_who::test_runlevel
    test_who::test_time

test result: FAILED. 2716 passed; 80 failed; 40 ignored; 0 measured; 0 filtered out; finished in 231.09s
```

### CODE REVIEW

File: src/uu/pinky/src/pinky.rs [hunk 1]
```diff
@@ -120,6 +141,17 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
     }
 }
 
+#[cfg(target_os = "openbsd")]
```
Comment (sylvestre) [inline src/uu/pinky/src/pinky.rs:144]: please add a comment explaining why it is unspported

File: src/uu/uptime/src/uptime.rs [hunk 2]
```diff
@@ -54,6 +61,15 @@ pub fn uumain(args: impl uucore::Args) -> UResult<()> {
     }
 }
 
+#[cfg(all(unix, target_os = "openbsd"))]
+#[uucore::main]
```
Comment (sylvestre) [inline src/uu/uptime/src/uptime.rs:65]: same

File: src/uu/pinky/src/pinky.rs [hunk 3]
```diff
@@ -5,19 +5,38 @@
 
 // spell-checker:ignore (ToDO) BUFSIZE gecos fullname, mesg iobuf
 
+#[cfg(not(target_os = "openbsd"))]
```
Comment (sylvestre) [inline src/uu/pinky/src/pinky.rs:8]: not a fan of this long list of openbsd exception either 
it is possible to have pinky-openbsd.rs and switch in the cargo.toml instead ?

File: src/uu/pinky/src/pinky.rs (hunk 1 above)
Comment (lcheylus) [reply src/uu/pinky/src/pinky.rs:144]: OK, I will add a comment to explain why it's unsupported on OpenBSD.

File: src/uu/pinky/src/pinky.rs (hunk 3 above)
Comment (lcheylus) [reply src/uu/pinky/src/pinky.rs:8]: I'm not a fan either of these duplicate cfg exceptions for `target_os="openbsd"`.
For the 4 unsupported tools (`pinky`, `who`, `users`, `uptime`), I propose to create an alternate implementation `<tool>_openbsd.rs` (with comments for explanation) and use it via conditional compilation in `Cargo.toml`.

Do you agree before I make my changes ?

Comment (sylvestre) [reply src/uu/pinky/src/pinky.rs:8]: let's do that :)

Comment (lcheylus) [reply src/uu/pinky/src/pinky.rs:8]: I have a problem to implement this solution : conditional compilation with `cfg` in not supported for `lib` in `Cargo.toml` (see https://github.com/rust-lang/cargo/issues/12260).

For example, for `uptime` tool, I have:
- original file `src/uu/uptime/src/uptime.rs`
- alternate implementation for OpenBSD in `src/uu/uptime/src/uptime_openbsd.rs`
- `lib` defined in `Cargo.toml` as:
```
[lib]
path = "src/uptime.rs"
```

How to build `uptime` with `src/uu/uptime/Cargo.toml` using OpenBSD implementation for `lib` ?

Comment (sylvestre) [reply src/uu/pinky/src/pinky.rs:8]: @tertsdiepraam any idea how to do it? :)

Comment (lcheylus) [reply src/uu/pinky/src/pinky.rs:8]: With a little help from StackOverflow, I found a solution :)

For example, for `uptime` tool (sources in `src/uu/uptime/`):
- keep original implementation `src/uptime.rs`
- alternate implementation for OpenBSD in `src/uptime_openbsd.rs` (tool "unsupported")
- in `Cargo.toml`, rename `uptime.rs` to `lib.rs` in `lib` section

- `lib.rs`:
```rust
#[cfg(not(target_os = "openbsd"))]
mod uptime;
#[cfg(not(target_os = "openbsd"))]
pub use uptime::*;

#[cfg(target_os = "openbsd")]
mod uptime_openbsd;
#[cfg(target_os = "openbsd")]
pub use uptime_openbsd::*;
```

Are you OK with this solution before I push the corresponding modifications in this PR ?

Comment (sylvestre) [reply src/uu/pinky/src/pinky.rs:8]: parfait, ship it ! :)

Comment (lcheylus) [reply src/uu/pinky/src/pinky.rs:8]: @tertsdiepraam Thanks for the tip with `platform` directory for alternate implementation for different OS (as in `src/uu/whoami/src/platform/`). Some functions (`uu_app`) and variables (`options`) must be shared between each OS.

I have modified my code for `pinky`, `who`, `users` and `uptime` tools to use `platform` directory and specific implementation for OpenBSD.

I am checking my builds and tests on OpenBSD and Linux before submitting a new version of my PR.

Comment (lcheylus) [reply src/uu/pinky/src/pinky.rs]: **New version of my PR** pushed (`git push -f`) to add OpenBSD specific implementation for pinky, users, uptime and who utils.

Thanks for the review and merge if OK.

File: src/uu/pinky/src/platform/openbsd.rs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+// Specific implementation for OpenBSD: tool unsupported (utmpx not supported)
+
+use crate::uu_app;
+
+use uucore::error::UResult;
+
+pub fn uumain(args: impl uucore::Args) -> UResult<()> {
+    let _matches = uu_app().try_get_matches_from(args)?;
+
+    println!("unsupported");
```
Comment (sylvestre) [inline src/uu/pinky/src/platform/openbsd.rs:15]:
```suggestion
    println!("unsupported command on OpenBSD");
```

Comment (sylvestre) [reply src/uu/pinky/src/platform/openbsd.rs:15]: and for other occurrences

Comment (lcheylus) [reply src/uu/pinky/src/platform/openbsd.rs:15]: OK, fix message ("unsupported command on OpenBSD") for `pinky`, `uptime`, `users` and `who`.

**Comment by sylvestre** [comment]
Looks great to me. @tertsdiepraam what do you think?

**Review by tertsdiepraam — COMMENTED** [review]
I'm honestly not really sure about this. I don't really like to introduce utils that compile but don't work at all. That's because if someone installs uutils, it usually overrides the standard implementation. If we have a stub for, say, `pinky`, then the standard `pinky` gets overridden, so can't be used easily anymore. So I think disabling them in `Cargo.toml` until we support them makes more sense. I do agree that this is the right folder structure though. I just think we should tackle it one at a time by implementing them.

Why don't we start with a feature flag like this:
```toml
openbsd = [
  "feat_Tier1",
  #
  "feat_require_crate_cpp",
  "feat_require_unix",
  "feat_require_unix_hostid",
]
```
Then we can add the utils as they are implemented (and once all of them are done we set it to `openbsd = ["unix"]`). Maybe our feature flags in general are overdue for a refactor, too.

Also, I think `other.rs` should be called `unix.rs` or `utmpx.rs` :)

The first four commits are definitely good though!

**Comment by lcheylus** [comment]
> Why don't we start with a feature flag like this:
> 
> ```toml
> openbsd = [
>   "feat_Tier1",
>   #
>   "feat_require_crate_cpp",
>   "feat_require_unix",
>   "feat_require_unix_hostid",
> ]
> ```
> 
> Then we can add the utils as they are implemented (and once all of them are done we set it to `openbsd = ["unix"]`). Maybe our feature flags in general are overdue for a refactor, too.

I have already proposed a PR with this solution, @sylvestre rejected it and asked me to implement as in this PR. See [PR#5613](https://github.com/uutils/coreutils/pull/5613)

My goal is to implement these 4 utils on OpenBSD, without using `utmpx`.

> Also, I think `other.rs` should be called `unix.rs` or `utmpx.rs` :)

OK, rename `platform/other.rs` => `platform/unix.rs` for the 4 utils.

**Comment by tertsdiepraam** [comment]
Alright, I talked with @sylvestre and we'll follow his plan! Thanks for working on this!

**Comment by lcheylus** [comment]
@sylvestre Build and tests OK on OpenBSD/amd64 with your 2 additional commits from https://github.com/lcheylus/rust-coreutils/commit/c32e730a1f0f0625158571dd528cde7519515f81

**Comment by sylvestre** [comment]
Super, merci :)

**Comment by sylvestre** [comment]
Félicitations!

**Pull request merged**